File: ai-proxy/laravel/app/Http/Controllers/OpenDocAiGatewayController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Client\Response as ClientResponse;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\RateLimiter;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\StreamedResponse;
use Throwable;

final class OpenDocAiGatewayController extends Controller
{
    public function health(): JsonResponse
    {
        return response()->json([
            'ok' => true,
            'authenticated' => config('opendoc-ai.token') !== '',
            'provider' => config('opendoc-ai.provider'),
            'model' => config('opendoc-ai.model'),
            'clientModelSelection' => (bool) config('opendoc-ai.allow_client_model'),
        ]);
    }

    public function options(Request $request): Response
    {
        if (!$this->originAllowed($request)) return $this->error('Origin is not allowed by this AI gateway.', 403);
        return $this->cors($request, response('', 204));
    }

    public function models(Request $request): JsonResponse
    {
        if ($failure = $this->guard($request)) return $failure;
        try {
            $models = $this->allowedModels();
            return $this->cors($request, response()->json([
                'models' => array_map(fn (string $model) => [
                    'id' => $model,
                    'label' => "{$model} · Gateway allowed",
                    'tier' => $this->tier($model),
                ], $models),
                'gateway' => array_filter([
                    'clientModelSelection' => (bool) config('opendoc-ai.allow_client_model'),
                    'provider' => config('opendoc-ai.provider'),
                    'model' => config('opendoc-ai.model'),
                    'models' => config('opendoc-ai.allow_client_model') ? $models : null,
                ], fn ($value) => $value !== null),
            ]));
        } finally {
            $this->releaseConcurrency();
        }
    }

    public function chat(Request $request): Response
    {
        if ($failure = $this->guard($request)) return $failure;
        try {
            if (strlen($request->getContent()) > config('opendoc-ai.max_body_bytes')) {
                $this->releaseConcurrency();
                return $this->error('AI gateway request body is too large.', 413);
            }
            $messages = $request->input('messages');
            if (!$this->validMessages($messages)) {
                $this->releaseConcurrency();
                return $this->error('The messages array exceeds gateway limits or is invalid.', 400);
            }
            $model = $this->selectModel($request);
            if ($model instanceof JsonResponse) {
                $this->releaseConcurrency();
                return $model;
            }
            if (!config('opendoc-ai.api_key') && config('opendoc-ai.provider') !== 'ollama') {
                $this->releaseConcurrency();
                return $this->error('AI_API_KEY is not configured on the gateway.', 503);
            }
            $headers = ['Content-Type' => 'application/json'];
            if (config('opendoc-ai.api_key')) $headers['Authorization'] = 'Bearer ' . config('opendoc-ai.api_key');
            if (config('opendoc-ai.site_url')) $headers['HTTP-Referer'] = config('opendoc-ai.site_url');
            if (config('opendoc-ai.app_name')) $headers['X-Title'] = config('opendoc-ai.app_name');
            $base = config('opendoc-ai.base_url');
            $url = str_ends_with($base, '/chat/completions') ? $base : $base . '/chat/completions';
            $temperature = is_numeric($request->input('temperature'))
                ? max(0, min(2, (float) $request->input('temperature')))
                : 0.2;
            $upstream = Http::withHeaders($headers)
                ->timeout(config('opendoc-ai.timeout_seconds'))
                ->connectTimeout(min(10, config('opendoc-ai.timeout_seconds')))
                ->withOptions(['stream' => true])
                ->post($url, [
                    'model' => $model,
                    'messages' => $messages,
                    'temperature' => $temperature,
                    'max_tokens' => config('opendoc-ai.max_output_tokens'),
                    'stream' => true,
                ]);
            if (!$upstream->successful()) {
                $this->releaseConcurrency();
                return $this->error($this->upstreamError($upstream), 502, [
                    'code' => 'upstream_error',
                    'status' => $upstream->status(),
                    'provider' => config('opendoc-ai.provider'),
                    'model' => $model,
                ]);
            }
            $psr = $upstream->toPsrResponse();
            $stream = $psr->getBody();
            $contentType = $upstream->header('Content-Type') ?: 'text/event-stream; charset=utf-8';
            $response = response()->stream(function () use ($stream): void {
                try {
                    while (!$stream->eof()) {
                        echo $stream->read(8192);
                        if (function_exists('ob_flush')) @ob_flush();
                        flush();
                        if (connection_aborted()) break;
                    }
                } finally {
                    $stream->close();
                    $this->releaseConcurrency();
                }
            }, 200, [
                'Content-Type' => $contentType,
                'Cache-Control' => 'no-cache, no-transform',
                'X-Accel-Buffering' => 'no',
            ]);
            return $this->cors($request, $response);
        } catch (Throwable $error) {
            $this->releaseConcurrency();
            report($error);
            return $this->error('AI gateway request failed.', 502);
        }
    }

    private function guard(Request $request): ?JsonResponse
    {
        $token = (string) config('opendoc-ai.token');
        if ($token === '' && !config('opendoc-ai.dev_mode'))
            return $this->error('AI_GATEWAY_TOKEN is required unless AI_GATEWAY_DEV_MODE=true.', 503);
        if ($token !== '' && !hash_equals('Bearer ' . $token, (string) $request->header('Authorization')))
            return $this->error('Invalid AI gateway token.', 401);
        if (!$this->originAllowed($request)) return $this->error('Origin is not allowed by this AI gateway.', 403);
        $rateKey = 'opendoc-ai:' . ($request->ip() ?: 'unknown');
        if (RateLimiter::tooManyAttempts($rateKey, config('opendoc-ai.rate_limit')))
            return $this->error('AI gateway rate limit exceeded.', 429);
        RateLimiter::hit($rateKey, 60);
        $active = Cache::increment('opendoc-ai:active');
        if ($active > config('opendoc-ai.max_concurrent')) {
            Cache::decrement('opendoc-ai:active');
            return $this->error('AI gateway is busy.', 429);
        }
        return null;
    }

    private function releaseConcurrency(): void
    {
        if ((int) Cache::get('opendoc-ai:active', 0) > 0) Cache::decrement('opendoc-ai:active');
    }

    private function originAllowed(Request $request): bool
    {
        $origin = (string) $request->header('Origin');
        return $origin === '' || in_array($origin, config('opendoc-ai.origins'), true);
    }

    private function cors(Request $request, Response $response): Response
    {
        $origin = (string) $request->header('Origin');
        if ($origin !== '') $response->headers->set('Access-Control-Allow-Origin', $origin);
        $response->headers->set('Vary', 'Origin');
        $response->headers->set('Access-Control-Allow-Headers', 'Content-Type, Authorization');
        $response->headers->set('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
        $response->headers->set('Access-Control-Expose-Headers', 'X-RateLimit-Limit, X-RateLimit-Remaining, Retry-After');
        return $response;
    }

    private function error(string $message, int $status, array $details = []): JsonResponse
    {
        return response()->json(['error' => ['message' => $message] + $details], $status);
    }

    private function allowedModels(): array
    {
        return config('opendoc-ai.allow_client_model')
            ? config('opendoc-ai.allowed_models')
            : [(string) config('opendoc-ai.model')];
    }

    private function selectModel(Request $request): string|JsonResponse
    {
        if ($request->filled('provider') && $request->input('provider') !== config('opendoc-ai.provider'))
            return $this->error("Provider is fixed to '" . config('opendoc-ai.provider') . "' by the gateway.", 400);
        $model = $request->filled('model') ? trim((string) $request->input('model')) : (string) config('opendoc-ai.model');
        return in_array($model, $this->allowedModels(), true)
            ? $model
            : $this->error("Model '{$model}' is not allowed by this gateway.", 400);
    }

    private function validMessages(mixed $messages): bool
    {
        if (!is_array($messages) || count($messages) < 1 || count($messages) > config('opendoc-ai.max_messages')) return false;
        $total = 0;
        foreach ($messages as $message) {
            if (!is_array($message)
                || !in_array($message['role'] ?? null, ['system', 'user', 'assistant'], true)
                || !is_string($message['content'] ?? null)
                || strlen($message['content']) > config('opendoc-ai.max_message_chars')) return false;
            $total += strlen($message['content']);
            if ($total > config('opendoc-ai.max_context_chars')) return false;
        }
        return true;
    }

    private function tier(string $model): string
    {
        return config('opendoc-ai.provider') === 'ollama' ? 'local' : (str_ends_with($model, ':free') ? 'free' : 'premium');
    }

    private function upstreamError(ClientResponse $response): string
    {
        $payload = $response->json();
        return is_array($payload)
            ? ($payload['error']['message'] ?? $payload['message'] ?? "Upstream returned HTTP {$response->status()}.")
            : "Upstream returned HTTP {$response->status()}.";
    }
}

File: ai-gateways/rails/app/controllers/open_doc_ai_gateway_controller.rb
require "net/http"
require "json"

class OpenDocAiGatewayController < ActionController::API
  include ActionController::Live

  before_action :apply_cors
  before_action :authorize_gateway, only: %i[models chat]
  before_action :take_limits, only: %i[models chat]
  after_action :release_non_streaming, only: :models

  def health
    render json: {
      ok: true,
      authenticated: OpenDocAiGateway::TOKEN.present?,
      provider: OpenDocAiGateway::PROVIDER,
      model: OpenDocAiGateway::MODEL,
      clientModelSelection: OpenDocAiGateway::ALLOW_CLIENT_MODEL
    }
  end

  def options
    head :no_content
  end

  def models
    render json: {
      models: OpenDocAiGateway::ALLOWED_MODELS.map { |model|
        { id: model, label: "#{model} · Gateway allowed", tier: model_tier(model) }
      },
      gateway: {
        clientModelSelection: OpenDocAiGateway::ALLOW_CLIENT_MODEL,
        provider: OpenDocAiGateway::PROVIDER,
        model: OpenDocAiGateway::MODEL,
        models: OpenDocAiGateway::ALLOW_CLIENT_MODEL ? OpenDocAiGateway::ALLOWED_MODELS.to_a : nil
      }.compact
    }
  end

  def chat
    unless request.content_length.to_i <= OpenDocAiGateway::MAX_BODY_BYTES
      return stream_error("AI gateway request body is too large.", 413)
    end
    messages = params[:messages]
    return stream_error("The messages array exceeds gateway limits or is invalid.", 400) unless valid_messages?(messages)
    if params[:provider].present? && params[:provider] != OpenDocAiGateway::PROVIDER
      return stream_error("Provider is fixed to '#{OpenDocAiGateway::PROVIDER}' by the gateway.", 400)
    end
    model = params[:model].presence || OpenDocAiGateway::MODEL
    return stream_error("Model '#{model}' is not allowed by this gateway.", 400) unless OpenDocAiGateway::ALLOWED_MODELS.include?(model)
    if OpenDocAiGateway::API_KEY.empty? && OpenDocAiGateway::PROVIDER != "ollama"
      return stream_error("AI_API_KEY is not configured on the gateway.", 503)
    end

    uri = URI(OpenDocAiGateway::BASE_URL.end_with?("/chat/completions") ? OpenDocAiGateway::BASE_URL : "#{OpenDocAiGateway::BASE_URL}/chat/completions")
    upstream_request = Net::HTTP::Post.new(uri)
    upstream_request["Content-Type"] = "application/json"
    upstream_request["Authorization"] = "Bearer #{OpenDocAiGateway::API_KEY}" if OpenDocAiGateway::API_KEY.present?
    upstream_request["HTTP-Referer"] = OpenDocAiGateway::SITE_URL if OpenDocAiGateway::SITE_URL.present?
    upstream_request["X-Title"] = OpenDocAiGateway::APP_NAME if OpenDocAiGateway::APP_NAME.present?
    temperature = params[:temperature].is_a?(Numeric) ? params[:temperature].clamp(0, 2) : 0.2
    upstream_request.body = JSON.generate(
      model: model,
      messages: messages.map { |message| { role: message[:role], content: message[:content] } },
      temperature: temperature,
      max_tokens: OpenDocAiGateway::MAX_OUTPUT_TOKENS,
      stream: true
    )

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == "https"
    http.open_timeout = [OpenDocAiGateway::UPSTREAM_TIMEOUT, 10].min
    http.read_timeout = OpenDocAiGateway::UPSTREAM_TIMEOUT
    http.write_timeout = OpenDocAiGateway::UPSTREAM_TIMEOUT

    http.request(upstream_request) do |upstream|
      if upstream.code.to_i.between?(200, 299)
        response.status = 200
        response.headers["Content-Type"] = upstream["Content-Type"].presence || "text/event-stream; charset=utf-8"
        response.headers["Cache-Control"] = "no-cache, no-transform"
        response.headers["X-Accel-Buffering"] = "no"
        upstream.read_body { |chunk| response.stream.write(chunk) }
      else
        raw = +""
        upstream.read_body { |chunk| raw << chunk if raw.bytesize < 16_384 }
        response.status = 502
        response.headers["Content-Type"] = "application/json; charset=utf-8"
        response.stream.write(JSON.generate(error: {
          message: upstream_message(raw, upstream.code.to_i),
          code: "upstream_error",
          status: upstream.code.to_i,
          provider: OpenDocAiGateway::PROVIDER,
          model: model
        }))
      end
    end
  rescue StandardError => error
    Rails.logger.error("OpenDoc AI gateway upstream error: #{error.class}")
    stream_error("AI gateway request failed.", 502) unless response.committed?
  ensure
    release_concurrency
    response.stream.close
  end

  private

  def apply_cors
    origin = request.headers["Origin"].to_s
    if origin.present? && !OpenDocAiGateway::ORIGINS.include?(origin)
      return render json: { error: { message: "Origin is not allowed by this AI gateway." } }, status: :forbidden
    end
    response.headers["Access-Control-Allow-Origin"] = origin if origin.present?
    response.headers["Vary"] = "Origin"
    response.headers["Access-Control-Allow-Headers"] = "Content-Type, Authorization"
    response.headers["Access-Control-Allow-Methods"] = "GET, POST, OPTIONS"
    response.headers["Access-Control-Expose-Headers"] = "X-RateLimit-Limit, X-RateLimit-Remaining, Retry-After"
  end

  def authorize_gateway
    return if OpenDocAiGateway::TOKEN.empty? && OpenDocAiGateway::DEV_MODE
    return if ActiveSupport::SecurityUtils.secure_compare(request.headers["Authorization"].to_s, "Bearer #{OpenDocAiGateway::TOKEN}")
    render json: { error: { message: "Invalid AI gateway token." } }, status: :unauthorized
  end

  def take_limits
    key = "opendoc-ai:rate:#{request.remote_ip}:#{Time.now.to_i / 60}"
    count = Rails.cache.increment(key, 1, expires_in: 70.seconds, initial: 0)
    if count > OpenDocAiGateway::RATE_LIMIT
      response.headers["Retry-After"] = "60"
      return render json: { error: { message: "AI gateway rate limit exceeded." } }, status: :too_many_requests
    end
    @concurrency_acquired = true
    active = Rails.cache.increment("opendoc-ai:active", 1, expires_in: 2.minutes, initial: 0)
    if active > OpenDocAiGateway::MAX_CONCURRENT
      release_concurrency
      response.headers["Retry-After"] = "2"
      return render json: { error: { message: "AI gateway is busy." } }, status: :too_many_requests
    end
    response.headers["X-RateLimit-Limit"] = OpenDocAiGateway::RATE_LIMIT.to_s
    response.headers["X-RateLimit-Remaining"] = [OpenDocAiGateway::RATE_LIMIT - count, 0].max.to_s
  end

  def release_non_streaming
    release_concurrency
  end

  def release_concurrency
    return unless @concurrency_acquired
    @concurrency_acquired = false
    Rails.cache.decrement("opendoc-ai:active", 1)
  end

  def valid_messages?(messages)
    return false unless messages.is_a?(Array) && messages.length.between?(1, OpenDocAiGateway::MAX_MESSAGES)
    total = 0
    messages.all? do |message|
      role = message[:role] || message["role"]
      content = message[:content] || message["content"]
      valid = %w[system user assistant].include?(role) && content.is_a?(String) && content.length <= OpenDocAiGateway::MAX_MESSAGE_CHARS
      total += content.length if valid
      valid && total <= OpenDocAiGateway::MAX_CONTEXT_CHARS
    end
  end

  def model_tier(model)
    return "local" if OpenDocAiGateway::PROVIDER == "ollama"
    model.end_with?(":free") ? "free" : "premium"
  end

  def upstream_message(raw, status)
    parsed = JSON.parse(raw)
    parsed.dig("error", "message") || parsed["message"] || "Upstream returned HTTP #{status}."
  rescue JSON::ParserError
    raw.presence || "Upstream returned HTTP #{status}."
  end

  def stream_error(message, status)
    response.status = status
    response.headers["Content-Type"] = "application/json; charset=utf-8"
    response.stream.write(JSON.generate(error: { message: message }))
  end
end
